cineast-core/src/main/java/org/vitrivr/cineast/core/iiif/imageapi/ImageRequestFactory.java

```diff
@@ -0,0 +1,404 @@
+  private final IIIFConfig iiifConfig;
+  private ImageApiVersion imageApiVersion;
+
+  public ImageRequestFactory(IIIFConfig iiifConfig) {
+    this.iiifConfig = iiifConfig;
+    try {
+      imageApiVersion = iiifConfig.getImageApiVersion();
+    } catch (IllegalArgumentException e) {
+      LOGGER.debug("No valid Image API version specified in the config file");
+      imageApiVersion = null;
+    }
+  }
+
+  private static boolean isParamStringValid(String configRegion) {
+    return configRegion != null && configRegion.length() != 0;
+  }
+
+  public List<ImageRequest> createImageRequests(String jobDirectoryString, String itemPrefixString) {
+    // If Image API version is not specified then auto-detect the highest API version supported by the server
+    if (imageApiVersion == null) {
+      LOGGER.debug("Image API version not specified. Starting detection of highest Image API version supported by the server.");
+      ImageApiVersion highestSupportedApiVersion = determineHighestSupportedApiVersion();
+      if (highestSupportedApiVersion == null) {
+        LOGGER.error("The Image API version could not be auto detected and thus Image Requests could not be created.");
+        return Collections.emptyList();
+      } else {
+        this.imageApiVersion = highestSupportedApiVersion;
+        LOGGER.debug("The highest supported Image API version was autodetected to version " + this.imageApiVersion.toNumericString());
+      }
+    }
+    if (imageApiVersion.equals(new ImageApiVersion(IMAGE_API_VERSION.TWO_POINT_ONE_POINT_ONE))) {
+      return new ApiJob_v2(iiifConfig).run(jobDirectoryString, itemPrefixString);
+    } else if (imageApiVersion.equals(new ImageApiVersion(IMAGE_API_VERSION.THREE_POINT_ZERO))) {
+      return new ApiJob_v3(iiifConfig).run(jobDirectoryString, itemPrefixString);
+    }
+    return new LinkedList<>();
+  }
+
+  /**
+   * Determines the highest version of the Image API supported by the server by making an Image Information request and trying to parse as different versions of the {@link ImageInformation} starting with the highest version {@link ImageInformation_v3} and moving downwards
+   */
+  @Nullable
+  public ImageApiVersion determineHighestSupportedApiVersion() {
+    List<IIIFItem> iiifItems = iiifConfig.getIiifItems();
+    if (iiifItems != null) {
+      Optional<IIIFItem> iiifItem = iiifItems.stream().findAny();
+      if (iiifItem.isPresent()) {
+        String identifier = iiifItem.get().getIdentifier();
+        String response = null;
+        String url = iiifConfig.getBaseUrl() + "/" + identifier;
+        try {
+          response = ImageInformationRequest.fetchImageInformation(url);
+        } catch (IOException e) {
+          LOGGER.error("An error occurred while making an Image Information request with the url: " + url);
+          e.printStackTrace();
+        }
+        if (response != null) {
+          try {
+            new ObjectMapper().readValue(response, ImageInformation_v3.class);
```

Comment:
Isn't there a way to look at the response directly, rather than using a deserialization exception?

Reply:
Deserilaization is a computationally cheap operation to perform and is only executed in the rare cases where the Image Api version is unknown. An alternative way would be to manually search for key-value substrings in the JSON but this method is prone to errors as there can be matching strings in the JSON that do not correspond to the data we are looking for and thus give a false negative.